### DISCUSSION TITLE
Approximation of inheritence

### DISCUSSION THREAD

**Discussion by simonsan** [rust-unofficial/patterns#315] [Ideas]
Reading this comment: https://www.reddit.com/r/rust/comments/w5hcbs/comment/ih7zuur/?context=3

it might be nice to write something up on that and link that to [`Deref` as an anti-pattern ](https://github.com/rust-unofficial/patterns/blob/main/anti_patterns/deref.md) as a "solution".

Q: So my question is what would be the most convenient way for a user to be able to add data to their structs that is safe (the user the can't delete it) and doesn't involve the user having to manually add the same data and functions over and over for each class.

A:
```rust
struct BaseData {}
trait DataObject: AsRef<BaseData> + AsMut<BaseData> {}
```

> if you require that types implement DataObject, then users will have to store a BaseData as one of their fields in order to satisfy the supertrait requirements, and now you can access your basic data without knowing anything else about what the user struct contains

**Comment by pickfire** [comment]
Is this an issue that we should open in clippy?

**Reply by simonsan** [reply]
Interesting idea, you mean to make `clippy` recognize the `deref` antipattern and give a recommendation, right?

**Reply by pickfire** [reply]
Yes
